### Introduction

For this homework, we will take ideas from the entire class: language models, text generation, vector-based word representations, syntactic analysis, and neural networks. We'll be using large, pre-trained language models to generate text, and studying how we can fine-tune these large language models to generate text in whatever genre and style we want!

In this assignment you will get:
1. An overview of the "Transformer" architecture is and why it is particularly well suited for Natural Language Processing tasks
2. An introduction to the Generative Pretrained Transformer (GPT) family, which is a set of large-scale language models that can be used to generate text that often sounds like it was written by a human.
3. Experience with using the HuggingFace package to fine-tune these models to generate text that sounds like it comes from a specific source.

# Problem One

# Part 1: What is a Transformer? (Reading)
<figure align="center">
<img src="https://media.giphy.com/media/VeWllmR9zfaco/giphy.gif" />
<figcaption>(It's probably not this guy, right?)</figcaption>
</figure>

### The Transformer

The current state-of-the-art for a variety of natural language processing tasks belongs to the **Transformer** architecture, first published December 6th 2017.

The Transformer can be thought of as a big feed-forward network with every feed-forward layer containing something called an "attention module".

>You might be wondering: why are we moving back to feed-forward networks after having so much success with recurrent neural networks and variants like LSTMs? Aren't RNNs naturally poised to handle sequences as their inputs? Well, as it turns out, the sequential nature of RNNs make them really difficult to train in a distributed/parallel fashion. So while RNNs make more sense to use on sequences of inputs, serial networks such as the transformer can be trained much faster, allowing orders of magnitude more training data to be used.



### Reading \# 1 - What is a Transformer?

In order to get a good grasp on exactly *why* these models are so good it's important to understand what they are and how they work.

Your first task for this homework is to read the blog post ["The Illustrated Transformer" by Jay Alammar](http://jalammar.github.io/illustrated-transformer/). This blog post explains the transformer architecture (and the all-important "Attention Module") with helpful visualizations and diagrams.

**You should read this post very closely and understand exactly what the Transformer is and how it works. Once you're finished reading, answer the following questions in 2-3 sentences each.**

1. (2 pts) What is Self-Attention (at a high level)?

   > Self-attention lies as the first element in the encoder and decoder of a transformer. The concept behind self attention is that, we use attention to process all the other words in the sentence besides the one that's being currently processed by the model. This helps in finding a better encoding of the current word.


2. (2 pts) How is Self-Attention computed?

   > First we create the query, key and value vectors from the encoders input vectors. Next, we calculate a score by taking the dot product of query and key vectors of each of the words in the input sentence. Next, we divide the score by 8(can be different) and pass the result through softmax to normalize the scores. Then, multiply each value vector by the normalized score. Finally, sum up the wighted value vectors.

3. (2 pts) What do the "Query", "Key", and "Value" vectors encode (at a high level)?

   > The "Query," "Key," and "Value" vectors play an important role in computing attention scores in transformer models. These vectors encode different aspects of the input word. The "Query" vector represents the information that the model seeks or queries from the input word(word that the model currently processes), the "Key" vector contains the information that helps determine the relevance of other words to the queried word, and the "Value" vector holds the actual content or information associated with the input word.

4. (2 pts) What is an attention "head" and why should we use multiple heads?

   > An attention "head" refers to a distinct set of Query, Key, and Value weight matrices. The Transformer model employs multiple attention heads, often eight in the example mentioned, which allows the model to focus on different aspects and positions within the input sequence of words. Using multiple heads enhances the model's capacity to capture diverse relationships and dependencies between the words in the input sequence.

5. (2 pts) What are positional embeddings?

   > Positional embeddings are additional vectors added to the input embeddings to account for the order or position of words in the input sequence. These embeddings follow a pattern, typically learned during the training process. The positional embeddings provide the model with a sense of the relative positions or distances between words in the input sequence. By incorporating these values into the input embeddings, the model can better capture the sequential information and understand the order of words.

6. (2 pts) Why are positional embeddings important?

   > Positional embeddings are important because they address the limitation of the self-attention mechanism in the Transformer model, which doesn't inherently capture word order in asentence. Without positional embeddings, the model would treat the input as an unordered set, which may end up in loss of vital information about the structure and meaning derived from word order. By adding positional embeddings to the input embeddings, the model gains the ability to distinguish the positions of words within the sequence.




-----------
### Reading \# 2 - Transformer Language Models

On June 11th 2018, OpenAI released a model named **GPT**, standing for *Generative Pre-Trained Transformer*. This model was a Transformer architecture that was modified such that it could be used for Text Generation instead of sequence to sequence modeling. This model was also pre-trained, which means that anyone could download the fully trained model and use it without needing to train the model themselves.

On February 14th 2019, OpenAI released a blog post detailing a brand new version of GPT that had an insane **1.5 billion parameters**. They named this version **GPT-2**. To train such a large model, OpenAI crawled 40GB worth of text from the web (roughly 20,000,000,000 words).

GPT-2 is an extremely impressive language model that can generate text that often sounds so plausible that it seems like it might have been written by a human.  Here is an example of what GPT-2 can generate, taken from [OpenAI’s blog post](https://openai.com/blog/better-language-models/):

**Human-Written Prompt:**
> “Miley Cyrus was caught shoplifting from Abercrombie and Fitch on Hollywood Boulevard today.”

**Model Continuation:**
> The 19-year-old singer was caught on camera being escorted out of the store by security guards.
>
> The singer was wearing a black hoodie with the label ‘Blurred Lines’ on the front and ‘Fashion Police’ on the back.
>
> Scroll down for video
>
> Shoplifting: Miley Cyrus was caught shoplifting from Abercrombie and Fitch on Hollywood Boulevard today (pictured)
>
> The singer was also wearing a pair of black-rimmed glasses, a black jacket, black jeans and black sandals.
>
> She was carrying a pair of black and white striped gloves and a small black bag.

*(To be clear: Miley Cyrus was never actually arrested for shoplifting)*.  

The quality of the model's output was so good, that the researchers at OpenAI were worried that, in addition to positive applications, it might be used for malicious purposes like generating fake news, impersonating people online, creating abusive or fake content to post on social media, or automating spam and phishing scams. In a [demo to Wired Magazine](https://www.wired.com/story/ai-text-generator-too-dangerous-to-make-public/), the reporter typed in “Hillary Clinton and George Soros” and GPT-2 generated a crazy rant with lots of conspiracy theories:
> Hillary Clinton and George Soros are a perfect match, and their agenda appears to be to create a political movement where Soros and his political machine and Clinton are two of the only major players. This is the first time Soros and Clinton have been caught on tape directly colluding in promoting the same false narrative. One of the key revelations in the leaked audio was Clinton's admission to a Russian banker that she knew about the Uranium One deal before it was approved by Congress. Clinton was shown sharing the same talking points that were originally drafted by a Fusion GPS contractor hired by an anti-Trump Republican donor. The leaked audio is the clearest evidence yet that the Clinton campaign and the Hillary Foundation colluded with Fusion GPS to manufacture propaganda against President Trump.


They were concerned enough that they labeled GPT-2 "too dangerous to release", and OpenAI initially refused to release their dataset, training code, or GPT-2 model weights.  OpenAI decided to release in a delayed, phased fashion so that researchers could spend time working on automatic detection of generated text.

In this homework, you'll get to be the judge of how good GPT-2 is, as you'll be using it yourself to generate text!

**To start your journey into the world of Text Generation, you should read Part 1 of the blog post ["The Illustrated GPT-2" by Jay Alammar](http://jalammar.github.io/illustrated-gpt2/) and answer the following questions in 2-3 sentences each**

7. (4 pts) How does the architecture of GPT-2 differ from the standard Encoder-Decoder Transformer model?
   > The transformer architecture first started off with a stack of encoder and decoder blocks with a connection from the encoder blocks to the decoder blocks. In GPT-2 however, we make use of only the decoder blocks. GPT-2 takes an input sequence and generates an output one token at a time. As the GPT-2 model generates an output token, it also adds this output token into the input sequence. This method is called auto regression.

8. (4 pts) What is the difference between "Masked Self-Attention" and "Self-Attention"
   > The key difference between "Masked Self-Attention" and "Self-Attention" lies in their attendance of future tokens during the attention calculation process. In a normal self-attention mechanism, as used in BERT, during the attention calculation we are allowed to attend to tokens both to its left and right, enabling it to gather information from the entire context. On the other hand, masked self-attention, intervenes in the self-attention calculation by blocking information from tokens located to the right of the position being processed.
   
9. (4 pts) What are logits? How are they computed? and How does GPT-2 use them to decide which word to predict next?
   > Logits are scores assigned to each word in the model's vocabulary, indicating the probability of each word being the next one in the sequence. In GPT-2, firstly the input sequence is passed through each of the decoder blocks self attention and feed foward part. Then, logits are computed by multiplying the output vector of the top decoder block in the model by the embedding matrix. Each row in the embedding matrix corresponds to the embedding of a word in the model's vocabulary. GPT uses a top_k to return the top k words that have the highest probability. A top_k = 40 gives the top 40 words with the highes probability and this technique achieves better results than top_k =1.

### Aside: GPT-3

On June 11th 2020, OpenAI released GPT-3 [(paper)](https://arxiv.org/pdf/2005.14165.pdf) [(wikipedia)](https://en.wikipedia.org/wiki/GPT-3). This model has an unfathomable **175 billion parameters** (100x larger than GPT-2!) and was trained on 570GB of text! This model is virtually indistinguishable from human output and can generate text about any topic and in any style with only a few words of priming text. It can do some very terrifying things.

GPT-3 Can:
- Generate JSX code off natural language descriptions
- Generate Emojis based off of descriptions of the feeling
- Generate regular expressions off natural language descriptions
- Generate website mockups off natural language descriptions
- Generate charts with titles, labels and legends from natural language descriptions
- Explain python code in plain english
- Automatically generate quiz questions (and grade them)
- Generate Latex from natural language descriptions
- Generate Linux commands from natural language descriptions
- Generate a Machine Learning model from natural language descriptions

[Here's a collection of 21 things GPT-3 can do (with examples)](https://machinelearningknowledge.ai/openai-gpt-3-demos-to-convince-you-that-ai-threat-is-real-or-is-it/#OpenAI_GPT-3_Demos)

[Here's a NYT article about how GPT-3 can write code, poetry, and argue](https://www.nytimes.com/2020/11/24/science/artificial-intelligence-ai-gpt3.html)

[Here's an article GPT-3 wrote for The Guardian about how it loves humans and would never subjugate humanity](https://www.theguardian.com/commentisfree/2020/sep/08/robot-wrote-this-article-gpt-3)

**You may optionally choose to read Jay Alammar's most recent blog post ["How GPT3 Works - Visualizations and Animations"](http://jalammar.github.io/how-gpt3-works-visualizations-animations/) from July 2020 if you're curious as to how GPT-3 differs from GPT-2**




-------------------------------

# Part 2: GPT-2 Text Generation with HuggingFace

Phew, that was a lot of reading. Now lets get to the fun part! Let's use the transformer to generate some text!!

We will use the [Transformers library from HuggingFace](https://transformer.huggingface.co), which provides support for many Transformer-based language models like GPT-2.

**IMPORTANT: Make sure that you have GPU set as your Hardware Accelerator in `Runtime > Change runtime type` before running this Colab.**

In [ ]:
!pip install transformers

## 2.1 The 'Pipeline' Interface

The simplest way to use the HuggingFace library is to use their [Pipeline interface](https://huggingface.co/transformers/main_classes/pipelines.html)

There are many different types of Pipelines available but in this section we'll use the TextGenerationPipeline to get up and running with pretrained gpt2 as fast as possible

In [ ]:
from transformers import pipeline

In [ ]:
# Note: device=0 means to use GPU, device=-1 is to use CPU
generator = pipeline('text-generation', model='gpt2', device=0)

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

In [ ]:
outputs = generator('I wonder what I will generate?')
print(outputs)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': "I wonder what I will generate? What if I only need a little help with math? It would be nice if that were the case. My problem is not going to be that the problem has to be solved; it just doesn't need to be"}]


Note that the 'text-generation' pipeline will work with any **auto-regressive** language model (a.k.a 'causal-lm' models according to the HuggingFace lingo). You can find a list of all such models here https://huggingface.co/models?filter=causal-lm.

10. (6 pts) **Your first task is to use the Pipeline interface to get generation output below for at least two different 'causal-lm' models (One of these two can be a different version of GPT2, but make sure at least one is a non-gpt family language model)**

In [ ]:
# Use a pipeline as a high-level helper
from transformers import pipeline

generator = pipeline("text-generation", model="EleutherAI/gpt-neo-125m")


In [ ]:

outputs = generator('I wonder what I will generate?')
print(outputs)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:1273: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(


[{'generated_text': 'I wonder what I will generate?\n\nI have a lot of questions about the future of the'}]


In [ ]:
generator = pipeline("text-generation", model="42dot/42dot_LLM-SFT-1.3B")

config.json:   0%|          | 0.00/694 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/5.76G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/144 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/753 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/870k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/529k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.26M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/1.27k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/1.23k [00:00<?, ?B/s]

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


In [ ]:

outputs = generator('I wonder what I will generate?')
print(outputs)

/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:1473: UserWarning: You have modified the pretrained model configuration to control generation. This is a deprecated strategy to control generation and will be removed soon, in a future version. Please use and modify the model generation configuration (see https://huggingface.co/docs/transformers/generation_strategies#default-text-generation-configuration )
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:1273: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(


[{'generated_text': 'I wonder what I will generate?\n\nA: You can use the `generate` method of'}]


## 2.2 Dissecting the Pipeline
Now that was easy!

As beautiful and easy as the Pipeline interface is, we want to know what's going on under the hood!

There are four main steps to a text generation pipeline:
1. (Tokenize) Turn the raw input text into a vector of integer token IDs using a tokenizer

2. (Encode) Feed those token IDs into the language model by querying for each token's embedding in the model's embedding matrix (the "encoder") and then feed the "encoded" sequence into the decoder module

3. (Decode) The decoder will output logits (a probability distribution over all possible integer token IDs) and we sample from those logits to get our next token -- repeat until EOS token is generated or we hit max_length

4. (Detokenize) Take the output sequence of token IDs and turn them from integer token IDs back to tokens with the tokenizer

Below you'll see how HuggingFace does this:

First we have to initialize both the tokenizer and the model from their pre-trained checkpoints. Note that the tokenizer has to match the model.

In [ ]:
from transformers import GPT2Tokenizer, GPT2LMHeadModel# AutoTokenizer, AutoModelForCausalLM

tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2LMHeadModel.from_pretrained('gpt2').cuda()

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

In [ ]:
#### Step 1: Tokenize the input into integer token IDs
inputs = tokenizer.encode("Hello, how are you?", return_tensors='pt').to(model.device)
print("Input Token IDs: " + str(inputs))

Input Token IDs: tensor([[15496,    11,   703,   389,   345,    30]], device='cuda:0')


In [ ]:
#### Step 2 and 3: Feed in the integer token IDs and get out a sequence of token IDs as output
outputs = model.generate(inputs)
print("Output Token IDs: " + str(outputs))

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Output Token IDs: tensor([[15496,    11,   703,   389,   345,    30,   198,   198,    40,  1101,
           257,  1310,  1643,   286,   257, 34712,    13,   314,  1101,   257]],
       device='cuda:0')


/usr/local/lib/python3.7/dist-packages/transformers/generation_utils.py:1364: UserWarning: Neither `max_length` nor `max_new_tokens` has been set, `max_length` will default to 20 (`self.config.max_length`). Controlling `max_length` via the config is deprecated and `max_length` will be removed from the config in v5 of Transformers -- we recommend using `max_new_tokens` to control the maximum length of the generation.
  UserWarning,


In [ ]:
#### Step 4: Feed in the integer token IDs and get out a sequence of token IDs as output
output_text = [tokenizer.decode(x) for x in outputs]
print("Output Text: " + str(output_text))

Output Text: ["Hello, how are you?\n\nI'm a little bit of a nerd. I'm a"]


Now that you have dissected the pipeline, it's time to play with some common parameters!

[Check out this demo notebook from HuggingFace](https://github.com/huggingface/blog/blob/master/notebooks/02_how_to_generate.ipynb) for a good overview of the different generation parameters and what they do (with example code!).

The full documentation on all of the parameters you can use in the generate function can be found [here](https://huggingface.co/transformers/main_classes/model.html#transformers.generation_utils.GenerationMixin.generate)

As an example, below we have a call to generate that:
- randomly samples from the top 50 words in the output distribution (rather than just greedily picking the best one every time)
- downweights the probability of all previously generated tokens by a factor of 1.2 (to prevent repetition)
- goes on for 512 tokens, because its more interesting

In [ ]:
inputs = tokenizer.encode("Hello, how are you?", return_tensors='pt').to(model.device)
outputs = model.generate(
      inputs,
      do_sample=True,          # Randomly sample from the logits instead of greedily picking next word with highest probability
      top_k=50,                 # Only sample from the top 50 most likely words
      repetition_penalty=1.2,    # Downweights the probability of all previously generated tokens by a factor of 1.2
      max_length=512          # Generate for a maximum of 512 tokens
  )
print([tokenizer.decode(x) for x in outputs][0])


The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Hello, how are you? What's wrong with me?"
I didn't even think I was mad about it. A look of fear and uncertainty had gripped my eyes like a flood on the desert floor as we stood here in front that long corridor separating us from some abandoned building surrounded by hundreds more tents… for what truly belonged to him—he is going crazy! My heart started racing against autopilot control every time he came into view when turning through his left-hand side or right angle (if still not blinking), "How am i hurting!?" He screamed at once across our face which caused all sorts pines within each footwork shift up high: pain coming out behind where they were pointing while blood ran everywhere throughout their bodies... no one saw them because people never let themselves fall so quickly without something else involved.. But there she is looking over her shoulder after everything happening around inside 'em?! Why would someone leave this place...? As soonAs anyone steps outside these tent area

**11. Your job is to provide two different examples of generation output from GPT-2 with different choices of generation parameters. You must also provide a 1-2 sentence explanation of what these parameters do and how they affect your output**

Feel free to get creative with this! Really poke around and try to find the combination of settings that gives you the best sounding text! The ways in which these parameters affect how 'human-like' a section of generated text sounds is an area of active research. :)

In [ ]:
## YOUR CODE HERE FOR HYPERPARAMETER VARIATION 1
inputs = tokenizer.encode("It's a beautiful sunny day, with a cute purple horizon", return_tensors='pt').to(model.device)
outputs = model.generate(
      inputs,
      max_length=100,
      num_beams=10,             # Find the top 10 most likely next word
      no_repeat_ngram_size=2,   # so that no 2-gram appears
      num_return_sequences=5,   # Return the top 5 sentences
      early_stopping=True       #generation is finished when all beam hypotheses reached the EOS token
  )

for i, beam_output in enumerate(outputs):
  print("{}: {}".format(i, tokenizer.decode(beam_output, skip_special_tokens=True)))
# print([tokenizer.decode(x) for x in outputs][0])

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


0: It's a beautiful sunny day, with a cute purple horizon, and the sun is shining through the trees.

I'm not sure if I'll be able to get out of the car, but I'm pretty sure it's going to be a long day. I don't know how long it will take me to find my way back to the house, so I hope it doesn't take too long.
1: It's a beautiful sunny day, with a cute purple horizon, and the sun is shining through the trees.

I'm not sure if I'll be able to get out of the car, but I'm pretty sure it's going to be a long day. I don't know how long it will take me to find my way back to the house, so I hope it doesn't take too long. It's been a while since I've had a chance to take a picture with my
2: It's a beautiful sunny day, with a cute purple horizon, and the sun is shining through the trees.

I'm not sure if I'll be able to get out of the car, but I'm pretty sure it's going to be a long day. I don't know how long it will take me to find my way back to the house, so I hope it doesn't take too long.

(4 pts) YOUR ANSWER HERE - EXPLANATION FOR HPARAM VARIATION 1



In the above implementation we make use of beam search with 10 beams to get the top 10 most likely probabilities, which provides more flexibility. no_repeat_ngram_size is a parameter that finds the most common n-grams penalty, which prevents an already occured n-gram from happening again. We are returning the top 5 sentences with the parameter 'num_return_sequences'.

In [ ]:
## YOUR CODE HERE FOR HYPERPARAMETER VARIATION 2

inputs = tokenizer.encode("Football is the best sport that we have created. A game that has all the adrenaline to make us jump ", return_tensors='pt').to(model.device)
outputs = model.generate(
      inputs,
      do_sample=True,          # Randomly sample from the logits instead of greedily picking next word with highest probability
      top_k=0,
      temperature=0.3,
      repetition_penalty=1.3,    # Downweights the probability of all previously generated tokens by a factor of 1.3
      max_length=256          # Generate for a maximum of 256 tokens
  )
print([tokenizer.decode(x) for x in outputs][0])


The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Football is the best sport that we have created. A game that has all the adrenaline to make us jump!"
"I think it's a great thing for our fans, and I'm sure they'll be excited about this."<|endoftext|>


(4 pts) YOUR ANSWER HERE -- EXPLANATION FOR HPARAM VARIATION 2


Here we use the sampling technique by setting do_sample to true and deactivating top_k by setting it to 0. 'repetition penalty' downweighs the probability of all previously generated tokens. Temperature is a parameter used to lower/soften the effect of softmax.

## 2.3 Fine-Tuning GPT-2
Okay now time for the best part!

Generating general-purpose text from pre-trained models is great, but what if we want our text to be in a specific genre or style? Luckily for us, the GPT family of models use the idea of "Transfer learning" -- using knowledge gained from one problem (or training setting), and applying it to another area or domain. The idea of transfer learning for NLP, is that we can train a language model on general texts, and then adapt it to use it for a specific task or domain that we're interested in. This process is also called **fine-tuning**.

In this section we'll walk you through an example of using HuggingFace to fine-tune GPT-2 and then you'll be asked to fine-tune GPT-2 on two datasets of your own choosing!

### Fine-Tuning Example using HuggingFace Datasets library: Crime and Punishment

For our fine-tuning example we're going to train GPT-2 to mimic the style of Fyodor Dostoevsky's novel "Crime and Punishment"

We will be downloading our data using the HuggingFace [Datasets](https://huggingface.co/docs/datasets/) library.

In [ ]:
!pip install datasets

In [ ]:
from transformers import Trainer, TrainingArguments, DataCollatorForLanguageModeling
import datasets
from datasets import load_dataset, list_datasets
from transformers import GPT2Tokenizer, GPT2LMHeadModel

### Step 1: Initialize a Brand New GPT-2 Model and Tokenizer

In [ ]:
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
tokenizer.pad_token = tokenizer.eos_token
model = GPT2LMHeadModel.from_pretrained('gpt2').cuda()
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)

###Step 2: Load the text of "Crime and Punishment" and tokenize it

The 'load_dataset' function queries for a dataset with a certain tag and downloads the corresponding data from HuggingFace's hosting site. This allows us to download all sorts of datasets through the same interface!

The documentation for load_dataset can be found [here](https://huggingface.co/docs/datasets/package_reference/loading_methods.html#datasets.load_dataset)

Here we take our tokenizer and run it on the entirety of Crime and Punishment in a single batch by using map on our custom encode function.

In [ ]:
def encode(batch): return tokenizer([x.strip('\n\r') for x in batch['line']], truncation=True, padding=True)

crime_and_punishment = load_dataset('crime_and_punish', split='train')
processed = crime_and_punishment.map(encode, batched=True, batch_size=len(crime_and_punishment))
processed.set_format('torch', columns=['input_ids', 'attention_mask'])

### Step 3: Initialize the Trainer

The 'Trainer' module is the main way we perform fine-tuning. In order to initialize a Trainer, you need a model, tokenizer, TrainingArguments, your training data (in a Dataset object) and something called a data_collator (which tells the Trainer not to look for a vector of labels).

In [ ]:
training_args = TrainingArguments(
    output_dir='/content/drive/MyDrive/Colab Notebooks/CS 505',
    overwrite_output_dir=True,
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=64,
    logging_steps=100,
    weight_decay=0.01,
    logging_dir='./logs',
)

trainer = Trainer(
    model=model,
    tokenizer=tokenizer,
    args=training_args,
    data_collator=data_collator,
    train_dataset=processed,
)

### Step 4: Fine-Tune the Model!

Now we're done! All we have to do is hit run and sit back!

In [ ]:
trainer.train()

TrainOutput(global_step=1374, training_loss=3.284515714020708, metrics={'train_runtime': 114.0319, 'train_samples_per_second': 192.657, 'train_steps_per_second': 12.049, 'total_flos': 396406039680000.0, 'train_loss': 3.284515714020708, 'epoch': 1.0})

### Step 5: Save the Model and use it to Generate!

Save your fine-tuned model and compare its output with regular GPT-2's output to see the difference for yourself!

In [ ]:
trainer.save_model('./dostoevskypt2')

In [ ]:
dostoevskypt2 = pipeline('text-generation', model='./dostoevskypt2', device=0)
gpt2 = pipeline('text-generation', model='gpt2', device=0)

In [ ]:
print(dostoevskypt2('Saint Petersburg is'))
print(gpt2('Saint Petersburg is'))

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'Saint Petersburg is haunted by the same morbid feeling of haunted illness: there are not many people who remember. “When the cold came in I couldn’t bring myself to go to the shop. They wouldn’t even take me out'}]
[{'generated_text': 'Saint Petersburg is an extremely important centre – the oldest city in Russia – one of the top five in the world and one of the second-best centers of science, technology, engineering, and mathematics in the world. It boasts an amazing population of nearly'}]


## PERPLEXITY

12. (2 pts) Using the pointer [here](https://huggingface.co/transformers/perplexity.html), compute the perplexity of the GPT2 pre-trained model on the Wikipedia test set (you can keep the same hyperparameters as in the link)

In [ ]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF GPT2 ON WIKIPEDIA TEST SET

# ANSWERS BELOW:
# Load wiki test set
from datasets import load_dataset
import torch
from tqdm import tqdm

test = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")
encodings = tokenizer("\n\n".join(test["text"]), return_tensors="pt")
max_length = model.config.n_positions
stride = 512

# Define a function for ppl
def ppl(model, input_ids_all, stride):
  nlls = []
  for i in tqdm(range(0, input_ids_all.size(1), stride)):
      begin_loc = max(i + stride - max_length, 0)
      end_loc = min(i + stride, input_ids_all.size(1))
      trg_len = end_loc - i  # may be different from stride on last loop
      input_ids = input_ids_all[:, begin_loc:end_loc].to("cuda:0")
      target_ids = input_ids.clone()
      target_ids[:, :-trg_len] = -100

      with torch.no_grad():
          outputs = model(input_ids, labels=target_ids)
          neg_log_likelihood = outputs[0] * trg_len

      nlls.append(neg_log_likelihood)

  ppl = torch.exp(torch.stack(nlls).sum() / end_loc)
  return ppl
ppl(model, encodings.input_ids, stride)


Token indices sequence length is longer than the specified maximum sequence length for this model (287644 > 1024). Running this sequence through the model will result in indexing errors
100%|██████████| 562/562 [00:20<00:00, 27.78it/s]


tensor(88.4710, device='cuda:0')

> YOUR PERPLEXITY ANSWER HERE:
Here we get a perplexity of 88.47 which seems to be low enough to show that the GPT-2 model generalises well enough, but we cannot just look at an absolute value of the perplexity and decide if the model is doing well or not. Instead we have to compare perplexity values across different models or fine-tuning scenarios to really understand how well it is doing.

13. (2 pts) Compute the  perplexity of the dostoevskypt2 model on Wikipedia test set




In [ ]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF DOSTOEVSKYPT2 ON WIKIPEDIA TEST SET

# ANSWERS BELOW:
from transformers import pipeline
from tqdm import tqdm
import torch

test_wikipedia = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")

tokenizer_dostoevsky = dostoevskypt2.tokenizer
encodings_wikipedia = tokenizer_dostoevsky("\n\n".join(test_wikipedia["text"]), return_tensors="pt")
max_length = dostoevskypt2.model.config.n_positions
stride = 512

ppl(dostoevskypt2.model, encodings_wikipedia.input_ids, stride)


100%|██████████| 562/562 [00:19<00:00, 29.30it/s]


tensor(68.7608, device='cuda:0')

> YOUR PERPLEXITY ANSWER HERE:
Here, our trained dostoevsky model seems to do better than GPT-2 by generalising better on the test wiki set. It has a lower perplexity score(68.76) than GPT-2(88.47)

14. (2 pts) Compute the perplexity of the GPT2 pre-trained model on the Crime and Punishment train dataset



In [ ]:
test = load_dataset('crime_and_punish', split='train')
encodings_crime_and_punishment = tokenizer("\n\n".join(test["line"]), return_tensors="pt")
max_length = model.config.n_positions
stride = 512
ppl(model, encodings_crime_and_punishment.input_ids, stride)



100%|██████████| 705/705 [00:25<00:00, 27.74it/s]


tensor(66.3355, device='cuda:0')

> YOUR PERPLEXITY ANSWER HERE
The pre-trained GPT-2 model generalises well on the crime and punishment train dataset with a perplexity of 66.33.

15. (2 pts) Compute the **train** perplexity of the **dostoevskypt2** model




In [ ]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF DOSTOEVSKYPT2 ON CRIME AND PUNISHMENT TRAIN DATASET
test = load_dataset('crime_and_punish', split='train')
encodings_crime_and_punishment = tokenizer("\n\n".join(test["line"]), return_tensors="pt")
max_length = dostoevskypt2.model.config.n_positions
stride = 512
ppl(dostoevskypt2.model, encodings_crime_and_punishment.input_ids, stride)

Token indices sequence length is longer than the specified maximum sequence length for this model (360591 > 1024). Running this sequence through the model will result in indexing errors
100%|██████████| 705/705 [00:23<00:00, 29.43it/s]


tensor(67.3069, device='cuda:0')

> YOUR PERPLEXITY ANSWER HERE
67.30


> (1 pt) Which model performs better on Crime and Punishment train set, vanilla GPT-2 or your dostoevskypt2 checkpoint?

> YOUR ANSWER HERE

Althought the perplexity does not seem to differ by a lot, GPT-2 does margginally better than dostoevskypt2 (66.4276 VS 67.3069)

16. (2 pts) Compute perplexity of the GPT2 model on your raw pride and prejudice text.

In [ ]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF GPT2 ON PRIDE AND PREJUDICE TEXT
with open("/content/drive/MyDrive/Colab Notebooks/CS 505/prideAndPrejudice.txt") as file:
    pride_and_prejudice_text = file.read()

encodings_pride_and_prejudice = tokenizer(pride_and_prejudice_text, return_tensors='pt')
ppl(model, encodings_pride_and_prejudice.input_ids, stride)



100%|██████████| 300/300 [00:10<00:00, 27.69it/s]


tensor(49.7871, device='cuda:0')

> YOUR PERPLEXITY ANSWER HERE
49.78

17. (2 pts) Compute perplexity of the **dostoevskypt2** model on your raw pride and prejudice text.

In [ ]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF dostoevskipt2 ON PRIDE AND PREJUDICE TEXT

encodings_pride_and_prejudice_dostoevskypt2 = tokenizer_dostoevsky(pride_and_prejudice_text, return_tensors='pt')
ppl(dostoevskypt2.model, encodings_pride_and_prejudice_dostoevskypt2.input_ids, stride)



100%|██████████| 300/300 [00:10<00:00, 29.40it/s]


tensor(42.6701, device='cuda:0')

> YOUR PERPLEXITY ANSWER HERE
42.67

### Now's Your Turn!

**Your job is to fine-tune GPT2 one more time with your choice of fine-tuning dataset.**

*****For the fine-tuned model you create, you should clearly demonstrate (through visible generation outputs and analysis) that your fine-tuned model follows the desired style better than vanilla GPT2** ***

Please make sure to give a brief description

In order to see which datasets are available for download, run the cell below. Pick one that you think would be interesting!

In [ ]:
from datasets import list_datasets
datasets_list = list_datasets()
print(', '.join(dataset for dataset in datasets_list))

### Tips
- Most of the datasets hosted by HuggingFace are not meant for Causal LM fine-tuning. Make sure you preprocess them accordingly if you want to use them.
- In order to check out information about a dataset hosted by huggingface you can use [this web viewer](https://huggingface.co/datasets/viewer/?dataset=crime_and_punish). Try to avoid downloading a dataset that's too big!
- You will likely have to change the custom 'encode' function for each new dataset you want to fine-tune on. You need to change batch['line'] to instead index with the correct column label for your specific dataset (it probably wont be called 'line').

### Useful Links
[load_datasets Documentation](https://huggingface.co/docs/datasets/package_reference/loading_methods.html#datasets.load_dataset)

[Trainer Documentation](https://huggingface.co/transformers/main_classes/trainer.html#id1)

[Example: Fine-Tuning BERT for Esperanto](https://colab.research.google.com/github/huggingface/blog/blob/master/notebooks/01_how_to_train.ipynb#scrollTo=zTgWPa9Dipk2)

[Example: Fine-Tuning for IMDb Classification](https://colab.research.google.com/drive/1-JIJlao4dI-Ilww_NnTc0rxtp-ymgDgM?usp=sharing#scrollTo=5DEWNilys9Ty)


#### 18. Dataset \#1

In [ ]:
## YOUR CODE HERE - FOR FINE-TUNING GPT2 ON DATASET
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
tokenizer.pad_token = tokenizer.eos_token
model = GPT2LMHeadModel.from_pretrained('gpt2').cuda()
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)

def encode(batch): return tokenizer([x.strip('\n\r') for x in batch['text']], truncation=True, padding=True)

emotion = load_dataset('dair-ai/emotion', split='train')
processed = emotion.map(encode, batched=True, batch_size=len(emotion))
processed.set_format('torch', columns=['input_ids', 'attention_mask'])


training_args = TrainingArguments(
    output_dir='/content/drive/MyDrive/Colab Notebooks/CS 505',
    overwrite_output_dir=True,
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=64,
    logging_steps=100,
    weight_decay=0.01,
    logging_dir='./logs',
)

trainer = Trainer(
    model=model,
    tokenizer=tokenizer,
    args=training_args,
    data_collator=data_collator,
    train_dataset=processed,
)


Extracting data files:   0%|          | 0/3 [00:00<?, ?it/s]

Generating train split:   0%|          | 0/16000 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/2000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/2000 [00:00<?, ? examples/s]

Map:   0%|          | 0/16000 [00:00<?, ? examples/s]

In [ ]:
trainer.train()

TrainOutput(global_step=1000, training_loss=4.1683763427734375, metrics={'train_runtime': 140.8344, 'train_samples_per_second': 113.609, 'train_steps_per_second': 7.101, 'total_flos': 702222336000000.0, 'train_loss': 4.1683763427734375, 'epoch': 1.0})

In [ ]:
trainer.save_model('./emotion')

(4 pts) YOUR ANSWER HERE - BRIEF DESCRIPTION OF THE DATASET YOU CHOSE

Here, we use the emotion dataset which is a dataset of English Twitter messages with six basic emotions: anger, fear, joy, love, sadness, and surprise. Now that we have fine tuned GPT-2 on the emotion dataset, let's see how it generates text related to emotions.

In [ ]:
emoion = pipeline('text-generation', model='./emotion', device=0)
gpt2 = pipeline('text-generation', model='gpt2', device=0)

print(emoion('I am so happy about'))
print(gpt2('I am so happy about'))

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'I am so happy about this pregnancy and feel like i have reached a milestone in my pregnancy journey that i know i have achieved a healthy level of sexual satisfaction i want to achieve and i am feeling incredibly blessed to have these wonderful people in front of me'}]
[{'generated_text': 'I am so happy about it. I like it. I\'m glad that the team came here to help us."\n\nThe Rockets are one of 24 teams to have won their first three playoff games under new owners, which have put them on pace'}]


We can see that the generation from our emotion model is more coherent than the one generated directly by the pre-trained GPT-2 model. This shows that we successfully achieved what we wanted out of our transfer learning task.

In [ ]:
## YOUR CODE HERE - FOR GENERATION WITH YOUR FINE-TUNED MODEL AND COMPARISON WITH REGULAR GPT2
# First we'll check the perplexity of pre-trained GPT-2 on the emotion train set
test = load_dataset('dair-ai/emotion', split='train')
encodings = tokenizer("\n\n".join(test["text"]), return_tensors="pt")
max_length = model.config.n_positions
stride = 512
ppl(model, encodings.input_ids, stride)

100%|██████████| 688/688 [00:24<00:00, 27.72it/s]


tensor(54.3119, device='cuda:0')

In [ ]:
# Now checking the perplexity of our fine tuned model
test = load_dataset('dair-ai/emotion', split='train')
encodings_emotion = emoion.tokenizer("\n\n".join(test["text"]), return_tensors="pt")
max_length = emoion.model.config.n_positions
stride = 512
ppl(emoion.model, encodings_emotion.input_ids, stride)

100%|██████████| 688/688 [00:23<00:00, 29.32it/s]


tensor(47.5005, device='cuda:0')

(5 pts) YOUR ANSWER HERE - COMPARISON OF YOUR DATASET'S FINE-TUNED OUTPUT VS NON-FINE-TUNED OUTPUT

In the code block above we calculate the perplexity scores of the pre-trained GPT-2 model on the emotions dataset. We recieve a perplexity of 54.3 which does not seem bad. Now, we calculate the perplexity scores of our fine-tuned model and see if it does better than the pre-trained GPT-2 model. We obtain a perplexity score of 47.5 and since lower perplexity score implies better results, we conclude that our fine-tuned model outperforms the pre-trained GPT-2 model.
